File: cpp/1110. 删点成林.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> ans;
    bool check(TreeNode* root,vector<int>& to_delete){
        for(auto v:to_delete){
            if(v==root->val){
                return true;
            }
        }
        return false;
    }
    TreeNode* dfs(TreeNode* root,vector<int>& to_delete,bool is_root){
        if(root==nullptr){
            return nullptr;
        }
        bool deleted=check(root,to_delete);
        root->left=dfs(root->left,to_delete,deleted);
        root->right=dfs(root->right,to_delete,deleted);
        
        if(deleted){
            root=nullptr;
        }else{
            if(is_root){
                ans.push_back(root);
            }
        }
        return root;
    }
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        dfs(root,to_delete,true);
        return ans;
    }
};

File: 刷题/golang/1110. 删点成林.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func delNodes(root *TreeNode, to_delete []int) []*TreeNode {
	toDeleted := make(map[int]bool)
	for _, val := range to_delete {
		toDeleted[val] = true
	}
	var ans []*TreeNode
	var dfs func(node *TreeNode, is_root bool) *TreeNode
	dfs = func(node *TreeNode, is_root bool) *TreeNode {
		if node == nil {
			return nil
		}
		deleted := false
		if toDeleted[node.Val] {
			deleted = true
		}
		node.Left = dfs(node.Left, deleted)
		node.Right = dfs(node.Right, deleted)
		if deleted {
			node = nil
		} else {
			if is_root {
				ans = append(ans, node)
			}
		}
		return node
	}
	dfs(root, true)
	return ans
}